src/ML/pytorch/valueMLmodel.cc

```diff
@@ -39,47 +40,21 @@ void valueMLmodel::learn(RLDataset& dataSet, unsigned int numberOfEpochs, unsign
 vector<float> valueMLmodel::calcValueEstimates(state* s) {
     /* Calculate all possible successor states of s.
        Returns a boolean vector corresponding to edges in such a way that true means an edge is selectable */
-    vector<bool> index = s->calcSuccessorStates();     
-
-    torch::Tensor succStates = vectorToTensor(s->successorStates);
-
-    // Returns the value prediction for all successor states.       
-    torch::Tensor pred = linearNet->forward(succStates);
-
-    vector<float> tempPrediction = tensorToVector(pred);
-    vector<float> prediction;
-    unsigned int counter = 0;
-
-    // To get a vector of the size of the state, we set 0 for every not selectable edge.
     for (unsigned int i = 0; i < dimstate; i++) {
-        if (index.at(i)) {
-            prediction.push_back(tempPrediction.at(counter));
-            counter++;
-        } else {
-        prediction.push_back(0);
-        }
-    }                                        
-    return prediction;                              
-}
-
-// Transforms a vector of states into a tensor
-torch::Tensor valueMLmodel::vectorToTensor(vector<vector<bool>>& vec) {
-    torch::Tensor tens = torch::zeros({(long int) vec.size(),(long int) vec.at(0).size()});
-    for (unsigned int i = 0; i < vec.size(); i++){
-        for (unsigned int j = 0; j < vec.at(i).size(); j++) {                
-            tens[i][j] = (float)vec.at(i).at(j);
-        }           
-    }
-    return tens;
-}
-
-// Transforms a tensor of values for the successor states into a vector
-vector<float> valueMLmodel::tensorToVector(torch::Tensor& tens) {
-    vector<float*> vec;
-    vector<float> vecReturn;
-    for (unsigned int i = 0; i < tens.numel(); i++) {
-        vec.push_back(tens[i].data<float>());
-        vecReturn.push_back(*vec.at(i));
+        tensState[i] = (float) s->selectedSubset.at(i);
+        
     }
-    return vecReturn;
+    vector<unsigned int> index = s->calcSuccessorStates();     
+    vector<float*> prediction(dimstate,0);
+    vector<float> prediction1(dimstate,0);
+    torch::Tensor pred = torch::zeros(dimstate);
+    for (unsigned int i = 0; i < index.size(); i++) {
+        tensState[index[i]] = 1;
+        
+        pred = linearNet->forward(tensState);
+        prediction.at(index.at(i)) = pred.data<float>();
+        prediction1.at(index.at(i)) = *prediction.at(index.at(i));
```

Comment:
Just copy the value from the 1x1 Tensor `pred` directly into `prediction1`.

Reply:
Also: I think the softmax is still missing? If you convert the predictions into a `std::vector<float>`, you won't be able to use torch's prebuild softmax function. You could concatenate the 1x1 Tensors `pred` to a (index.size())x1 Tensor and then just call `torch::nn::functional::softmax`. This would be probably more efficient then implementing the softmax yourself.